feat: add NIM FE (num_request_max) + runtime config metrics with periodic polling

keivenchang opened the pull request:
#### Overview:

Add runtime configuration metrics for model deployment monitoring across all backend frameworks. Fix missing max_num_seqs metrics in SGLang and TensorRT-LLM backends while maintaining existing vLLM functionality.

#### Details:

- Add periodic polling system for model runtime config metrics (max_num_seqs, total_kv_blocks, max_num_batched_tokens)
- **vLLM Backend**: Maintain existing functionality where max_num_seqs is retrieved from engine.vllm_config.scheduler_config
- **TensorRT-LLM Backend**: Add missing runtime config metric population by setting max_num_seqs from config.max_batch_size and max_num_batched_tokens from config.max_num_tokens
- **SGLang Backend**: Fix max_num_seqs retrieval by accessing server_args instead of scheduler_info due to SGLang's architectural separation of configuration parameters from runtime statistics
- Rename MODEL_WORKER_COUNT to MODEL_WORKERS_TOTAL following Prometheus conventions
- Add configuration safety guard against invalid polling intervals
- Update metrics documentation with new runtime configuration metrics

#### Where should the reviewer start?

- lib/llm/src/http/service/metrics.rs - Core metrics implementation with periodic polling
- components/backends/trtllm/src/dynamo/trtllm/main.py - TensorRT-LLM metric consistency improvements
- components/backends/sglang/src/dynamo/sglang/register.py - SGLang architectural fix for config vs runtime separation

#### Related Issues:

DIS-607 DIS-641

/coderabbit profile chill

keivenchang committed:
feat: add runtime config metrics with periodic polling

- Add new Prometheus metrics for model runtime configuration
- Include metrics for total_kv_blocks, max_num_seqs, max_num_batched_tokens
- Add MDC metrics for context_length, kv_cache_block_size, migration_limit
- Implement model health status tracking (healthy/unhealthy/unknown)
- Add background polling task to keep metrics current as backends change
- Preserve all historical data - metrics are never removed, only marked unhealthy
- Configurable poll interval via DYN_RUNTIME_CONFIG_METRICS_POLL_INTERVAL_SECS

Signed-off-by: Keiven Chang <keivenchang@users.noreply.github.com>

keivenchang commented on lib/llm/src/http/service/metrics.rs:
```diff
@@ -330,9 +428,200 @@ impl Metrics {
         Ok(())
     }
 
+    /// Update runtime configuration metrics for a model
+    /// This should be called when model runtime configuration is available or updated
+    pub fn update_runtime_config_metrics(
+        &self,
+        model_name: &str,
+        runtime_config: &ModelRuntimeConfig,
+    ) {
+        if let Some(total_kv_blocks) = runtime_config.total_kv_blocks {
+            self.model_total_kv_blocks
+                .with_label_values(&[model_name])
+                .set(total_kv_blocks as i64);
+        }
+
+        if let Some(max_num_seqs) = runtime_config.max_num_seqs {
+            self.model_max_num_seqs
+                .with_label_values(&[model_name])
+                .set(max_num_seqs as i64);
+        }
+
+        if let Some(max_batched_tokens) = runtime_config.max_num_batched_tokens {
+            self.model_max_num_batched_tokens
+                .with_label_values(&[model_name])
+                .set(max_batched_tokens as i64);
+        }
+    }
```
Updated to u64.

keivenchang commented on lib/llm/src/http/service/metrics.rs:
```diff
@@ -330,9 +428,200 @@
+            .set(migration_limit as i64);
+    }
+
+    /// Set model health status
+    /// 1 = healthy (model is active and responding)
+    /// 0 = unhealthy (model is known but not responding)
+    /// -1 = unknown (model status is unclear)
+    pub fn set_model_health_status(&self, model_name: &str, healthy: bool) {
+        let status = if healthy { 1 } else { 0 };
+        self.model_health_status
+            .with_label_values(&[model_name])
+            .set(status);
+    }
+
+    /// Mark model as having unknown health status
+    pub fn set_model_health_unknown(&self, model_name: &str) {
+        self.model_health_status
+            .with_label_values(&[model_name])
+            .set(-1);
+    }
+
```
Disagree. Explicit is better than implicit.

keivenchang committed:
refactor: rename MODEL_WORKER_COUNT to MODEL_WORKERS_TOTAL following Prometheus conventions

- Update constant name from MODEL_WORKER_COUNT to MODEL_WORKERS_TOTAL
- Change metric name from model_worker_count to model_workers_total
- Update all references in metrics service implementation
- Update documentation in README to reflect new metric name
- Improve metrics polling task lifecycle management in service_v2
- Follows Prometheus naming convention using _total suffix for counters

Signed-off-by: Keiven Chang <keivenchang@users.noreply.github.com>

keivenchang committed:
fix: guard against zero or negative DYN_HTTP_SVC_CONFIG_METRICS_POLL_INTERVAL_SECS

- Add validation to ensure poll interval is greater than 0
- Use filter() to reject zero or negative values
- Update documentation to clarify validation requirement
- Prevents potential issues with zero-duration polling intervals

Signed-off-by: Keiven Chang <keivenchang@users.noreply.github.com>

kthui commented on deploy/metrics/README.md:
```diff
@@ -79,6 +79,27 @@ When using Dynamo HTTP Frontend (`--framework VLLM` or `--framework TRTLLM`), th
+
+The frontend also exposes model configuration metrics with the `dynamo_frontend_model_*` prefix. These metrics are populated from the worker backend registration service when workers register with the system:
+
+**Runtime Config Metrics (from ModelRuntimeConfig):**
+These metrics come from the runtime configuration provided by worker backends during registration.
+
+- `dynamo_frontend_model_total_kv_blocks`: Total KV blocks available for a worker serving the model (gauge)
+- `dynamo_frontend_model_max_num_seqs`: Maximum number of sequences for a worker serving the model (gauge)
+- `dynamo_frontend_model_max_num_batched_tokens`: Maximum number of batched tokens for a worker serving the model (gauge)
+
+**MDC Metrics (from ModelDeploymentCard):**
+These metrics come from the Model Deployment Card information provided by worker backends during registration.
+
+- `dynamo_frontend_model_context_length`: Maximum context length for a worker serving the model (gauge)
+- `dynamo_frontend_model_kv_cache_block_size`: KV cache block size for a worker serving the model (gauge)
+- `dynamo_frontend_model_migration_limit`: Request migration limit for a worker serving the model (gauge)
```
I wonder if the metrics from a particular worker can be selected, or are the metrics aggregated from all registered workers?

kthui reviewed (commented):
I think it would be nice to have some extra test case(s) on the new metrics - to get a better understanding on how the new metrics would look from end users.

ryan-lempka commented on lib/llm/src/http/service/metrics.rs:
```diff
@@ -330,9 +438,240 @@ impl Metrics {
+                        .set(0);
+                }
+
+                // Note: If multiple model instances have the same name, only the first instance's config metrics are recorded.
```
When multiple workers share the same model name, can the selection be deterministic still somehow instead of first seen?

ryan-lempka reviewed (approved):
Few minor comments (non-blocking). Otherwise LGTM.

keivenchang committed:
docs: improve metrics documentation and fix naming

- Fix metric name from model_workers_total to model_workers
- Document model name deduplication behavior in README.md
- Add comments explaining gauge vs counter usage for runtime config metrics
- Clarify that some metrics use gauges because they're synchronized from upstream

Signed-off-by: Keiven Chang <keivenchang@users.noreply.github.com>

keivenchang committed:
test: add MDC registration integration test and fix fractional polling intervals

Signed-off-by: Keiven Chang <keivenchang@users.noreply.github.com>

keivenchang committed:
fix: add missing max_num_seqs metrics for SGLang and TensorRT-LLM backends

SGLang: get max_num_seqs from server_args since SGLang separates config from runtime stats
TensorRT-LLM: populate max_num_seqs and max_num_batched_tokens from config for metrics consistency

KrishnanPrash reviewed (approved):
LGTM! Left a note to document future work.

Note: Under the assumption that MDC are static, we can use a watcher pattern used by the frontend [[reference](https://github.com/ai-dynamo/dynamo/blob/f9be2e9eeab5862292429c70dfb2fae4c56334a1/lib/llm/src/entrypoint/input/http.rs#L207-L229)] instead of the polling approach currently being utilized.
